### ISSUE TITLE
Vertaisarviointi / peer-review

### ISSUE THREAD

**Issue by vainiovesa** [ileskaa/neural-network#1]
## Vertaisarviointi

Neuroverkon koodi on kirjoitettu tehokkaasti ja selkeästi. Käyttöliittymä on hauska ja havainnollistava, sekä yksikkötestit kattavia. Tässä vielä muutama parannusehdotus.

#### Vahvistusdata
Neuroverkon koulutuksen aikana kannattaa käyttää koulutusdatasta erillistä vahvistusdataa. Näin neuroverkon oppimista pystyy seuraamaan ja ehkäisemään ylisovittamista.

#### Neuroverkon koulutus käyttöliittymässä
Käyttäjän kannalta olisi hyvä päästä kokeilemaan erikokoisia neuroverkkoja ja kouluttamaan niitä eri oppimisnopeuksilla ja erikokoisilla minisatseilla. Tämä olisi toki järkevää toteuttaa niin, että koulutus tapahtuisi lokaalisti.

#### Muut testit
Testeihin kannattaa lisätä yksikkötestien lisäksi muita testejä (Omassa projektissani huomasin invarianttitestit hyödyllisiksi neuroverkon testaamiseen). Erityisesti testi, joka varmistaa, että neuroverkko ylisovittuu pieneen datan osajoukkoon varmistaa gradienttimenetelmän toimivuuden, kuten [Björkqvist](https://www.sebastianbjorkqvist.com/blog/writing-automated-tests-for-neural-networks/) neuvoo.

#### Lisädatan luominen
Neuroverkon tarkkuutta voi parantaa myös luomalla lisädataa olemassaolevasta koulutusdatasta. Kuvien määrän voi moninkertaistaa lisäämällä niihin kohinaa, liikuttelemalla piirrettyjä numeroita reunoille ja venyttämällä niitä.

#### Muuta pientä
* Pitäisikö [tässä](https://github.com/ileskaa/neural-network/blob/a6d7f1cc10e87d114cd889ea902091358207e6da/src/model/mlp.py#L226) olla max eikä mean? On mahdollista antaa syötteeksi kuva, jossa kaikki arvot eivät ole normalisoituja, kun keskiarvo on vain riittävän pieni.
* [Tässä](https://github.com/ileskaa/neural-network/blob/a6d7f1cc10e87d114cd889ea902091358207e6da/src/model/mlp.py#L46) "weighted sums" lienee kuvaavampi.
* Mahdollisuuden käyttää muita aktivointi-, alustus- ja tappiofunktioita voisi myös lisätä.

Kaiken kaikkiaan loistavaa työtä.

Kloonasin repositorion 4.8.2025, kello 12:27.

## Peer review

The code for the neural network is efficient and clear. The user interface is fun and illustrative, and the unit tests comprehensive. Here's a few suggestions for improvements yet.

#### Validation data
While training a neural network it makes sense to use validation data separate from the training data. It's then possible to monitor the learning process and prevent overfitting.

#### Training the network in the user interface
From the user's point of view it would be nice to try different size neural networks and train them with different learning rates and different size minibatches. Of course this should be done so that the training happens  locally.

#### Other tests
In addition to unit tests, other tests should be added (In my own project I found invariant tests useful for testing neural networks). In particular, a test making sure the network overfits to a small subset of the data verifies the functioning of the gradient descent algorithm as [Björkqvist](https://www.sebastianbjorkqvist.com/blog/writing-automated-tests-for-neural-networks/) advices.

#### Creating additional data
The accuracy of the neural network can be improved by creating additional data from the existing training data. The amount of the pictures can be multiplied by adding noise to them, moving the drawn digits to the edges and stretching them.

#### Other small things
* Should [here](https://github.com/ileskaa/neural-network/blob/a6d7f1cc10e87d114cd889ea902091358207e6da/src/model/mlp.py#L226) be max instead of mean? It's possible to give a picture as an input where all the values are not normalized when the mean is small enough.
* [Here](https://github.com/ileskaa/neural-network/blob/a6d7f1cc10e87d114cd889ea902091358207e6da/src/model/mlp.py#L46) "weighted sums" might be more descriptive.
* The possibility to use other activation, initialization and loss functions could be added as well.

All in all excellent work.

Repository cloned 4.8.2025, at 12:27.
